### PR TITLE
Improve application flow

### PR DESCRIPTION
**Pull request by joaoburatto**
# Summary

Marked as WIP because I need to test more. And prob document some stuff.

Improves the application initialization flow by giving us more control over the order which things are executed.

Specifically, this changes the application initialization flow from direct calls to other classes initialization methods to an event-based approach, classes that will be initialized need to listen to those events. This is done to reduce cluttering methods in one class, and to prevent this class to hold references for everything in the game, it can lead to major issues when adding/working on that class in the future, it was already painful to separate them at this time when it was simple.

Highlight to the logger initialization change, as Nath thought it was better to include it in a different PR, but I think the nature of this PR was also fix the logger issue where it wouldn't fire on application start, only when a connection to the server was made. This was causing logs not being logged, now logs will log when they need to log. Logs will log.

#### PR checklist
- [ ] The game builds properly without errors.

- [ ] No unrelated changes are present.

- [ ] No "trash" files are committed.

- [ ] Relevant code is documented.

- [ ] Update the related GitBook document, or create a new one if needed.

# Testing

The reviewer can test this PR by seeing if the logs are being logged since application start. The reviewer can also test this PR by seeing if all the initialization flux that the user sees is correctly processed. Such as using the launcher and using the .bat files to launch the game with command line args.

# Changes

This reworks how the application is initialized, the main changes are under ApplicationInitializerSystem, which now fires some events during the initialization phase, they can be used to initialize specific parts of the application in specific orders.

I also had to add scenes to the build settings, they weren't there for some reason.

I had to move a bunch of stuff around to make the assemblies work together, overall this improved how the scripts are separated and cleaned up a bunch of weird out-of-place scripts.

Many changes are related to renaming/namespace changes, those couldn't be avoided.

I also fixed a case where the Coimbra framework wouldn't be downloaded again if you deleted the library folder, out of place, I know, but I had to do it because I had to delete library to test a fix with the assembly stuff, it was broken and I was trying stuff around.

### CODE REVIEW

File: Assets/Scripts/SS3D/Application/Events/ApplicationInitialized.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿using Coimbra.Services.Events;
+
+namespace SS3D.Application.Events
+{
+	public partial struct ApplicationInitialized : IEvent { }
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Application/Events/ApplicationInitialized.cs:8]: doc

File: Assets/Scripts/SS3D/Application/Events/ApplicationInitializing.cs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+﻿using Coimbra.Services.Events;
+
+namespace SS3D.Application.Events
+{
+	public partial struct ApplicationInitializing : IEvent { }
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Application/Events/ApplicationInitializing.cs:8]: doc

File: Assets/Scripts/SS3D/Application/Events/ApplicationPreInitializing.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿using Coimbra.Services.Events;
+
+namespace SS3D.Application.Events
+{
+    public partial struct ApplicationPreInitializing : IEvent
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Application/Events/ApplicationPreInitializing.cs:8]: doc

File: Assets/Scripts/SS3D/Data/AssetsInitializationTrigger.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
+﻿using Coimbra.Services.Events;
+using SS3D.Application.Events;
+using SS3D.Core.Behaviours;
+using SS3D.Logging;
+
+namespace SS3D.Data
+{
+	public class AssetsInitializationTrigger : Actor
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Data/AssetsInitializationTrigger.cs:11]: doc

File: Assets/Scripts/SS3D/Launcher/LauncherSystem.cs [hunk 5]
```diff
@@ -21,18 +24,15 @@ public sealed class LauncherSystem : Core.Behaviours.System
 		/// <param name="port">The port number to use.</param>
 		public void LaunchGame(NetworkType networkType, string ckey, string ip, string port)
 		{
-			if (!Application.isEditor)
-			{
-				NetworkSettings.ResetOnBuiltApplication();
-			}
-
-			NetworkSettings networkSettings = ScriptableSettings.GetOrFind<NetworkSettings>();
+            NetworkSettings networkSettings = ScriptableSettings.GetOrFind<NetworkSettings>();
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Launcher/LauncherSystem.cs:25]: fix indent

File: Assets/Scripts/SS3D/Launcher/LauncherSystem.cs [hunk 6]
```diff
@@ -21,18 +24,15 @@ public sealed class LauncherSystem : Core.Behaviours.System
...
 
 			networkSettings.NetworkType = networkType;
 			networkSettings.Ckey = ckey;
 			networkSettings.ServerAddress = ip;
 			networkSettings.ServerPort = Convert.ToUInt16(port);
 
+            Log.Information(this, $"Launching game with on {networkType} for {ip} {port} as {ckey}", Logs.Important);
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Launcher/LauncherSystem.cs:32]: fix indent

File: Assets/Scripts/SS3D/Networking/Debug/BandwitdhView.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+﻿using Coimbra;
+using FishNet.Component.Utility;
+using SS3D.Networking.Settings;
+
+namespace SS3D.Networking.Debug
+{
+	public class BandwitdhView : BandwidthDisplay
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Networking/Debug/BandwitdhView.cs:10]: doc

File: Assets/Scripts/SS3D/Networking/Events/NetworkConnectionStarted.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿using Coimbra.Services.Events;
+
+namespace SS3D.Networking.Events
+{
+    public partial struct NetworkConnectionStarted : IEvent
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Networking/Events/NetworkConnectionStarted.cs]: doc

File: Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
+using FishNet.Object;
+using SS3D.Networking.Events;
+
+namespace SS3D.Networking
+{
+    /// <summary>
+    /// Simply calls network connection started events when a network connection is stablished.
+    /// </summary>
+    public class NetworkConnectionStartedListener : NetworkBehaviour
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs]: Not sure if this is a good idea, isn't there the possibility that things are set up for this script network wise but not on scripts listening to this event ?

File: Assets/Scripts/SS3D/Networking/Settings/NetworkSettings.cs [hunk 10]
```diff
@@ -40,6 +40,12 @@ public sealed class NetworkSettings : ScriptableSettings
 		/// </summary>
 		public ushort ServerPort = 2222;
 
+		/// <summary>
+		/// Enables Fish-net's bandwidth UI in the game.
+		/// </summary>
+		[Header("Debug Settings")]
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Networking/Settings/NetworkSettings.cs:46]: that is not related to the PR, remove please

**Review by stilnat — CHANGES_REQUESTED** [review]
Partial review, some doc and some questions

File: Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs (hunk 9 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs]: I don't understand what you mean by that

File: Assets/Scripts/SS3D/Networking/Settings/NetworkSettings.cs (hunk 10 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Networking/Settings/NetworkSettings.cs:46]: Its a very small addition, I see no harm adding it here

**Comment by stilnat** [comment]
the crafting loading bar does not show up when crafting

File: Assets/Scripts/SS3D/Networking/Settings/NetworkSettings.cs (hunk 10 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Networking/Settings/NetworkSettings.cs:46]: fine, I prefer having different features in different PR, but that one is very small so, fine

File: Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs (hunk 9 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs]: Script A register to NetworkConnectionStartedListener  NetworkConnectionStarted event.
-> OnStartClient is called on NetworkConnectionStartedListener  , which invoke  NetworkConnectionStarted event
-> script A trigger a method registered to the event.

Is there any guarantee Script A is correctly set up for network stuff ? Is the Owner, or observers set up ? Can I call any RPC ?

File: Assets/Scripts/SS3D/Application/Events/ApplicationInitialized.cs (hunk 1 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Application/Events/ApplicationInitialized.cs:8]: done

File: Assets/Scripts/SS3D/Application/Events/ApplicationInitializing.cs (hunk 2 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Application/Events/ApplicationInitializing.cs:8]: done

File: Assets/Scripts/SS3D/Application/Events/ApplicationPreInitializing.cs (hunk 3 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Application/Events/ApplicationPreInitializing.cs:8]: done

File: Assets/Scripts/SS3D/Data/AssetsInitializationTrigger.cs (hunk 4 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Data/AssetsInitializationTrigger.cs:11]: done

File: Assets/Scripts/SS3D/Launcher/LauncherSystem.cs (hunk 5 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Launcher/LauncherSystem.cs:25]: done I think

File: Assets/Scripts/SS3D/Launcher/LauncherSystem.cs (hunk 6 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Launcher/LauncherSystem.cs:32]: done I think

File: Assets/Scripts/SS3D/Networking/Debug/BandwitdhView.cs (hunk 7 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Networking/Debug/BandwitdhView.cs:10]: done

File: Assets/Scripts/SS3D/Networking/Events/NetworkConnectionStarted.cs (hunk 8 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Networking/Events/NetworkConnectionStarted.cs]: done

File: Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs (hunk 9 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs]: Added some guarantee for that.

File: Assets/Scripts/SS3D/Networking/Settings/NetworkSettings.cs (hunk 10 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Networking/Settings/NetworkSettings.cs:46]: k

File: Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs (hunk 9 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Networking/NetworkConnectionStartedListener.cs]: I don't see when this will be useful. It still feels somewhat dangerous to me. Since it's not used anywhere I rather have that removed for now.

File: Assets/Scripts/SS3D/SceneManagement/SceneSystem.cs [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+using SS3D.Application.Events;
+using SS3D.Data.AssetDatabases;
+using SS3D.Data.Generated;
+using SS3D.Logging;
+using System;
+
+// ReSharper disable ConditionIsAlwaysTrueOrFalse
+namespace SS3D.SceneManagement
+{
+	public sealed class SceneSystem : Core.Behaviours.System
```
Comment (stilnat) [inline Assets/Scripts/SS3D/SceneManagement/SceneSystem.cs:16]: Please explain a bit more in doc the current and future purpose of this system.

File: Assets/Scripts/SS3D/Networking/Events/NetworkConnectionStarted.cs [hunk 12]
```diff
@@ -0,0 +1,9 @@
+﻿using Coimbra.Services.Events;
+
+namespace SS3D.Networking.Events
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Networking/Events/NetworkConnectionStarted.cs]: might remove this one as well

**Review by stilnat — CHANGES_REQUESTED** [review]
can't build.  Apart from the changes below, please check that the logs still works for client/server/host, you'll find relevant info in PR's #1085 description. Also the map still appear broken too me, but not required to be fixed in this PR

File: Assets/Scripts/SS3D/Networking/Events/NetworkConnectionStarted.cs (hunk 12 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Networking/Events/NetworkConnectionStarted.cs]: Not sure why. This is a way to guarantee things will run only after the server started, if we don't want to use the NetworkBehaviour class.

File: Assets/Scripts/SS3D/SceneManagement/SceneSystem.cs (hunk 11 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/SceneManagement/SceneSystem.cs:16]: Done

**Comment by joaoburatto** [comment]
> can't build. Apart from the changes below, please check that the logs still works for client/server/host, you'll find relevant info in PR's #1085 description. Also the map still appear broken too me, but not required to be fixed in this PR

Fixed it. Not the map though.

**Review by stilnat — APPROVED** [review]
works as expected and no added bugs found, good work.
Also close #1292
